src/TokenHandler.mo

```diff
@@ -0,0 +1,364 @@
+
+    let map : Map = Map();
+    let backlog : BackLog = BackLog();
+
+    // a backlog of principals, waiting for consolidation
+    // var backlog : AssocList.AssocList<Principal, ()> = null;
+    // var backlogSize : Nat = 0;
+
+    /// query the usable balance
+    public func balance(principal : Principal) : Nat = info(principal).usable_balance;
+
+    /// query all tracked balances for debug purposes
+    public func info(principal : Principal) : Info and {
+      var usable_balance : Nat;
+    } {
+      let ?item = map.get(principal) else return {
+        var deposit = 0;
+        var credit = 0;
+        var usable_balance = 0;
+      };
+
+      let usableDeposit = Int.max(0, item.deposit - fee);
+      if (item.credit + usableDeposit < 0) {
+        Debug.trap("item.credit + Int.max(0, item.deposit - fee) < 0");
+      };
+
+      {
+        var deposit = item.deposit;
+        var credit = item.credit;
+        var usable_balance = Int.abs(item.credit + usableDeposit);
+      };
+    };
+
+    /// retrieve the current size of consolidation backlog
+    public func backlogSize() : Nat = backlog.size();
+
+    /// deduct amount from P’s usable balance. Return false if the balance is insufficient.
+    public func debit(principal : Principal, amount : Nat) : Bool {
+      map.change(
+        principal,
+        func(info) {
+          if (info.deposit + info.credit < amount) return false;
+          info.credit -= amount;
+          return true;
+        },
+      );
+    };
+
+    ///  add amount to P’s usable balance (the credit is created out of thin air)
+    public func credit(principal : Principal, amount : Nat) : () {
+      ignore map.set(
+        principal,
+        func(info) {
+          info.credit += amount;
+          true;
+        },
+      );
+    };
+
+    /// The handler will call icrc1_balance(S:P) to query the balance. It will detect if it has increased compared
+    /// to the last balance seen. If it has increased then it will adjust the deposit (and hence the usable_balance).
+    /// It will also schedule or trigger a “consolidation”, i.e. moving the newly deposited funds from S:P to S:0.
+    public func notify(principal : Principal) : async* () {
+      if (not map.lock(principal)) return;
+      try {
+        let latestBalance = await* loadBalance(principal);
+        updateDeposit(principal, latestBalance);
+        if (latestBalance > fee) {
+          // schedule consolidation for this principal
+          backlog.push(principal);
+          ignore processBacklog();
+        };
+        map.unlock(principal);
+      } catch err {
+        map.unlock(principal);
+        throw err;
+      };
+    };
+
+    /// process first account from backlog
+    public func processBacklog() : async () = async {
+      func consolidate(principal : Principal) : async () {
+        let latestBalance = await* loadBalance(principal);
```

Comment:
If this call fails (return with an Error) then we don't catch the Error. The Error will be passed through so `consolidate` will return with the same Error. Two problem:
- we don't release the lock
- we don't put the principal back into backlog

Reply:
Btw, it would be ok to have `try { }` span both calls (balance and transfer). That's allowed and is ok if the calls are a sequence and the second should not happen if the first one fails.

Reply:
fixed